zkvm/src/signature/prover.rs

```diff
@@ -0,0 +1,187 @@
+                r_i,
+                R_i,
+            },
+            precommitment,
+        )
+    }
+
+    pub fn receive_precommitments(
+        self,
+        nonce_precommitments: Vec<NoncePrecommitment>,
+    ) -> (PartyAwaitingCommitments, NonceCommitment) {
+        // Store received nonce precommitments in next state
+        (
+            PartyAwaitingCommitments {
+                nonce_precommitments,
+                transcript: self.transcript,
+                multikey: self.multikey,
+                x_i: self.x_i,
+                r_i: self.r_i,
+            },
+            self.R_i,
+        )
+    }
+}
+
+impl PartyAwaitingCommitments {
+    pub fn receive_commitments(
+        mut self,
+        nonce_commitments: Vec<NonceCommitment>,
+    ) -> Result<(PartyAwaitingSiglets, Scalar), VMError> {
+        // Check stored precommitments against received commitments
+        for (pre_comm, comm) in self
+            .nonce_precommitments
+            .iter()
+            .zip(nonce_commitments.iter())
+        {
+            // Make H(comm) = H(R_i)
+            let actual_precomm = comm.precommit();
+
+            // Compare H(comm) with pre_comm, they should be equal
+            let equal = pre_comm.0.ct_eq(&actual_precomm.0);
+            if equal.unwrap_u8() == 0 {
+                return Err(VMError::InconsistentWitness);
+            }
+        }
+
+        // Make R = sum_i(R_i). nonce_commitments = R_i from all the parties.
+        let R: RistrettoPoint = nonce_commitments.iter().map(|R_i| R_i.0).sum();
+
+        // Make c = H(X, R, m)
+        // The message `m` should already have been fed into the transcript.
+        let c = {
+            self.transcript
+                .commit_point(b"P", &self.multikey.aggregated_key().0);
+            self.transcript.commit_point(b"R", &R.compress());
+            self.transcript.challenge_scalar(b"c")
+        };
+
+        // Make a_i = H(L, X_i)
+        let X_i = VerificationKey((self.x_i * RISTRETTO_BASEPOINT_POINT).compress());
+        let a_i = self.multikey.factor_for_key(&X_i);
+
+        // Generate share: s_i = r_i + c * a_i * x_i
+        let s_i = self.r_i + c * a_i * self.x_i;
+
+        // Store received nonce commitments in next state
+        Ok((
+            PartyAwaitingSiglets {
+                multikey: self.multikey,
+                nonce_commitments,
+                c,
+                R,
+            },
+            s_i,
+        ))
+    }
+}
+
+impl PartyAwaitingSiglets {
+    pub fn receive_trusted_shares(self, shares: Vec<Scalar>) -> Signature {
+        // s = sum(s_i), s_i = shares[i]
+        let s: Scalar = shares.iter().map(|share| share).sum();
+        Signature { s, R: self.R }
+    }
+
+    pub fn receive_shares(
+        self,
+        shares: Vec<Scalar>,
+        pubkeys: Vec<VerificationKey>,
+    ) -> Result<Signature, VMError> {
+        // Check that all shares are valid
+        for (i, s_i) in shares.iter().enumerate() {
```

Comment:
i wonder if we need to represent shares as a state machine themselves:

```
SharePrecommitted -> H(R_i)
ShareCommitted -> R_i (if verified to have correct R_i)
ShareSigned -> (R_i,s_i) (if verified to have correct s_i)
```

Reply:
then we can store identifiers inside shares and match them up correctly, or at least, factor verification logic out into more localized pieces, so we don't have big loops over multiple arrays.

Reply:
we can do this in another PR as a follow-up, since this one seems to be correct and working

Reply:
So, curently we have:
```
NoncePrecommitment -> H(R_i)
NonceCommitment -> R_i (if verified to have correct R_i)
```
and we also used to have `Siglet -> R_i` (but you suggested removing it).

These structs seems to be the same thing as what you suggested above (slightly different names). So what your'e basically suggesting is that we store indexes in each of these structs as well, so we can match up each party with their index with their messages? (Then, would we also want to assign an index to each party?)

Reply:
Except the idea is to focus on the _state of the counter-party_ and organize necessary checks around it, not on individual components - nonces, precommitments, siglets etc. Those are just pieces of the "counterparty's state".

I think my suggested `Share*` naming is off and it needs to be more like `CounterpartyPrecommitted`, `CounterpartyCommitted`, `CounterpartySigned`. We can later see what's more convenient/efficient - treating oneself as counterparty too and repeat all messages and checks, or only track truly others' state.

Reply:
Maybe we need to start with another state: `Counterparty`, which holds compressed pubkey `X_i` which we can use as an identifier, and no longer worry about indices.

Reply:
I think I understand what you mean, but just to gain clarity on details, can we hop on a call to talk about this design today (after stand-up)?

Reply:
sure! also: if you prefer, we can merge this as is and iterate in smaller PRs on top.

Reply:
the reason we have state machine (`Counterparty*`) within a state machine (`Party*`) is because each party is simultaneously a "dealer". So in RP MPC we had a separate dealer, so each each party was worrying only about its own state, not the others'.

Reply:
Incorporated into the writeup!